### PR TITLE
refactor(framework) Add `get_fab` implementations for FAB delivery

### PR DESCRIPTION
**Pull request by charlesbvll**
### Checklist

- [ ] Implement proposed change
- [ ] Write tests
- [ ] Update [documentation](https://flower.ai/docs/writing-documentation.html)
- [ ] Make CI checks pass
- [ ] Ping maintainers on [Slack](https://flower.ai/join-slack/) (channel `#contributions`)

### CODE REVIEW

File: src/py/flwr/server/app.py [hunk 1]
```diff
@@ -74,7 +74,7 @@
 ADDRESS_FLEET_API_REST = "0.0.0.0:9093"
 
 DATABASE = ":flwr-in-memory-state:"
-BASE_DIR = get_flwr_dir() / "superlink" / "ffs"
+BASE_DIR = get_flwr_dir() / "ffs"
```
Comment (danieljanes) [inline src/py/flwr/server/app.py:77]:
```suggestion
BASE_DIR = get_flwr_dir() / "superlink" / "ffs"
```

File: src/py/flwr/server/app.py [hunk 2]
```diff
@@ -214,8 +214,8 @@ def run_superlink() -> None:
     # Initialize StateFactory
     state_factory = StateFactory(args.database)
 
-    # Initialize FfsFactory
-    ffs_factory = FfsFactory(args.storage_dir)
+    # Initialize StateFactory
```
Comment (danieljanes) [inline src/py/flwr/server/app.py:217]:
```suggestion
    # Initialize FfsFactory
```

File: src/py/flwr/server/app.py [hunk 3]
```diff
@@ -214,8 +214,8 @@ def run_superlink() -> None:
...
+    ffs_factory = FfsFactory(args.base_dir)
```
Comment (danieljanes) [inline src/py/flwr/server/app.py:218]:
```suggestion
    ffs_factory = FfsFactory(args.storage_dir)
```

File: src/py/flwr/server/app.py [hunk 4]
```diff
@@ -618,8 +621,8 @@ def _add_args_common(parser: argparse.ArgumentParser) -> None:
         default=DATABASE,
     )
     parser.add_argument(
-        "--storage-dir",
-        help="The base directory to store the objects for the Flower File System.",
+        "--base-dir",
+        help="The base directory to store the objects.",
```
Comment (danieljanes) [inline src/py/flwr/server/app.py:625]:
```suggestion
        "--storage-dir",
        help="The base directory to store Flower File System objects.",
```

File: src/py/flwr/server/superlink/driver/driver_servicer.py [hunk 5]
```diff
@@ -174,14 +175,21 @@ def GetRun(
...
+                fab_hash=run.fab_hash,
             )
         )
 
     def GetFab(
         self, request: GetFabRequest, context: grpc.ServicerContext
     ) -> GetFabResponse:
-        """Will be implemented later."""
-        raise NotImplementedError
+        """Get run information."""
```
Comment (danieljanes) [inline src/py/flwr/server/superlink/driver/driver_servicer.py:185]: c&p?

File: src/py/flwr/server/superlink/fleet/message_handler/message_handler.py [hunk 6]
```diff
@@ -124,5 +127,14 @@ def get_run(
...
+            fab_hash=run.fab_hash,
         )
     )
+
+
+def get_fab(
+    request: GetFabRequest, ffs: Ffs  # pylint: disable=W0613
+) -> GetFabResponse:
+    """Get FAB."""
+    fab = Fab(request.hash_str, ffs.get(request.hash_str)[0])
```
Comment (danieljanes) [inline src/py/flwr/server/superlink/fleet/message_handler/message_handler.py:139]: What happens when this fails?

Comment (charlesbvll) [reply src/py/flwr/server/superlink/fleet/message_handler/message_handler.py]: @danieljanes I updated the code to make the `get` function of `Ffs` return an optional. I can move this change to another PR if you prefer

File: src/py/flwr/server/superlink/driver/driver_servicer.py [hunk 7]
```diff
@@ -174,14 +175,22 @@ def GetRun(
...
-        raise NotImplementedError
+        """Get FAB from Ffs."""
+        log(DEBUG, "DriverServicer.GetFab")
+
+        ffs: Ffs = self.ffs_factory.ffs()
+        if result := ffs.get(request.hash_str):
+            fab = Fab(request.hash_str, result[0])
+            return GetFabResponse(fab=fab_to_proto(fab))
+
+        return GetFabResponse()
```
Comment (danieljanes) [inline src/py/flwr/server/superlink/driver/driver_servicer.py:193]: We should probably use a gRPC error code here

File: src/py/flwr/server/superlink/fleet/message_handler/message_handler.py [hunk 8]
```diff
@@ -124,5 +127,17 @@ def get_run(
...
+
+def get_fab(
+    request: GetFabRequest, ffs: Ffs  # pylint: disable=W0613
+) -> GetFabResponse:
+    """Get FAB."""
+    if result := ffs.get(request.hash_str):
+        fab = Fab(request.hash_str, result[0])
+        return GetFabResponse(fab=fab_to_proto(fab))
+
+    return GetFabResponse()
```
Comment (danieljanes) [inline src/py/flwr/server/superlink/fleet/message_handler/message_handler.py:143]: Should we start to return an error code alongside the response from `message_handler` functions?

File: src/py/flwr/server/superlink/driver/driver_servicer.py (hunk 7 above)
Comment (charlesbvll) [reply src/py/flwr/server/superlink/driver/driver_servicer.py:193]: @danieljanes I tried to replicate what we do in other functions (like `GetRun`). Should I open a separate PR to also modify `GetRun`'s behaviour?

**Review by danieljanes — APPROVED** [review]

**Pull request merged**